### ISSUE TITLE
feat: make intermediateFileSystem overwritable for file-system cache testing

### ISSUE THREAD

**Issue by privatenumber** [webpack/webpack#13372]
## Feature request
For Webpack's Compiler's `intermediateFileSystem` to be overwritable for writing tests that involve the file-system cache.

**What is the expected behavior?**
For Webpack to use the passed in filesystem when overwriting it:
```js
const fs = require('fs');
const { Volume } = require('memfs');

const mfs = new Volume();
const compiler = webpack(config);

compiler.inputFileSystem = ufs.use(fs).use(mfs);
compiler.outputFileSystem = mfs;
compiler.intermediateFileSystem = mfs;

compiler.run((error, stats) => {
	...
});
```
**What is motivation or use case for adding/changing the behavior?**
I am fixing a plugin bug that involves the WP5 file-system cache, and to test, I am plugging in a `memfs` Volume into `compiler.intermediateFileSystem`. I run two builds and reuse the first memfs Volume in the second build.

Currently, seems like fs is bound to `PackFileCacheStrategy` on the `webpack(config)` call, so there is no opportunity to overwrite it like there is to overwrite other fs properties like `compiler.outputFileSystem`.

**How should this be implemented in your opinion?**
Perhaps delay `PackFileCacheStrategy` instantiation until the compiler starts to run.

**Are you willing to work on this yourself?**
yes

**Comment by alexander-akait** [comment]
Why do not use `compiler.intermediateFileSystem = mfs;`?

**Comment by privatenumber** [comment]
That's what I'm doing (code snippet provided above).

`PackFileCacheStrategy` initializes before I can set it though, so by the time I set it, it's too late.

**Comment by alexander-akait** [comment]
What do you need test in cache?

**Comment by alexander-akait** [comment]
You can use `compiler.cache.hooks.get.tap`, `compiler.cache.hooks.store.tap` and `compiler.cache.hooks.shutdown.tap` to test, just set `stage` https://github.com/webpack/webpack/blob/master/lib/Cache.js#L159

**Comment by privatenumber** [comment]
I'm writing a [localization plugin](https://github.com/privatenumber/webpack-localize-assets-plugin/blob/develop/src/index.ts), which efficiently handles multiple locales by inserting string placeholders and swaps them out with localized strings for each locale.

The plugin parses the source to find instances of `__('string')` and replaces them with a placeholder. It breaks when Webpack 5 cache kicks in because the cache had the placeholders instead of `__('string')`.

I'm currently refactoring it to be stateless, and now I'm writing tests to ensure it works with WP5 filesystem caching.

I'm not necessarily trying to read/access/manipulate the cache. Just want to create a test environment where Webpack uses the filesystem cache.

**Comment by alexander-akait** [comment]
You can set `cache: { type: 'filesystem' }` run 2 or 3 or more times and check output of assets, we are doing it for testing, it is better because it is integration test, rely only on `intermediateFileSystem` is not good place to tests, because other plugin/loader/tools/etc can change many things

**Comment by privatenumber** [comment]
That's what I'm currently doing, with a [temporary cache directory](https://github.com/privatenumber/webpack-localize-assets-plugin/blob/0c88dd4ac7225cf056d788ee58173e08eb4f9ef4/tests/webpack-localize-assets-plugin.spec.ts#L608-L620).

I'm not blocked by this, but I would still like for `compiler.intermediateFileSystem` to be usable and overwritable the same way other compiler file-systems are. I think the API would be much more intuitive and flexible that way.

PS Really appreciate the fast support btw @alexander-akait Thank you!

**Comment by alexander-akait** [comment]
Also you can override `outputFileSystem` on `memory-fs` and read file contents from memory, it is faster, because you still need run 2/3 and more times, overriding `compiler.intermediateFileSystem` will not good place for testing

**Comment by privatenumber** [comment]
All my tests are [in-memory](https://github.com/privatenumber/webpack-localize-assets-plugin/blob/0c88dd4ac7225cf056d788ee58173e08eb4f9ef4/tests/utils.ts#L57-L58) 😉

Seems WP5 file-system cache doesn't read from `outputFileSystem` though.

**Comment by privatenumber** [comment]
I investigated and found that I can mutate the Compiler before `PackFileCacheStrategy` initializes via a plugin.

`createCompiler` [runs all plugins](https://github.com/webpack/webpack/blob/068ce839478317b54927d533f6fa4713cb6834da/lib/webpack.js#L69-L77) before it applies the default options and returns the `compiler`.

I wrote a very simple plugin to mutate the file-systems at the right time:
```ts
type CompilerOptions = Partial<webpack.Compiler>;

class ConfigureCompilerPlugin {
	options: CompilerOptions;
	constructor(options: CompilerOptions) {
		this.options = options;
	}

	apply(compiler: webpack.Compiler) {
		Object.assign(compiler, this.options);
	}
}
```

which I use like this:
```ts
...,
plugins: [
	new ConfigureCompilerPlugin({
		inputFileSystem: memfs,
		outputFileSystem: memfs,
		intermediateFileSystem: memfs,
	}),
],
```

Hope this helps anyone encountering the same challenge.

There seems to be a few more roadblocks I have to overcome with the cache configuration to make it work, but I think that's outside the scope of this issue.

Will be closing this now.

Thank you again for the support @alexander-akait

**Comment by dmnsgn** [comment]
> All my tests are [in-memory](https://github.com/privatenumber/webpack-localize-assets-plugin/blob/0c88dd4ac7225cf056d788ee58173e08eb4f9ef4/tests/utils.ts#L57-L58) 😉
> 
> Seems WP5 file-system cache doesn't read from `outputFileSystem` though.

Does that also mean that any added `buildDependencies` from a in-memory outputFileSystem (eg. `memfs`) will not be cache invalidated?

I have got a webpack config with:
- an entry point that is in memfs (`compiler.outputFileSystem = memoryFs`)
```js
      entry: {
        index: './index.js',
      },
```
- `cache: {  type: 'filesystem' },`
- `webpack-dev-server` and `webpack-dev-middleware`
```js
    app.use(
      webpackDevMiddleware(compiler, {
        outputFileSystem: compiler.outputFileSystem,
      })
    )
```

When I manually regenerate the "index.js" and writeFileSync it memfs, changes are written to the in memory file. But when I reload my app in browser, the file doesn't show any of the changes. Setting `cache: false,` work as expected though which would indicate an error.

If setting:
```js
        buildDependencies: {
          theEntryPoint: [pathToMy.entry],
        },
```

this warning seem to confirm that:
```console
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve '/Users/me/Projects/path/index.js' in 'Users/me/Projects/path'
<w> while resolving 'Users/me/Projects/path/index.js' in Users/me/Projects/path as file
<w>  at resolve commonjs Users/me/Projects/path/index.js
```

Let me know if this should this be a separate issue?

**Comment by privatenumber** [comment]
WP5 fs caching reads from `intermediateFileSystem`, so make sure you update that too via my solution [here](https://github.com/webpack/webpack/issues/13372#issuecomment-841598405)

**Comment by dmnsgn** [comment]
Yeah unfortunately even when setting `compiler.intermediateFileSystem = memoryFs`, it doesn't seem to pickup the fact that my in-memory `index.js` entry point has changed. The change btw, is writting a new `import * as Module from "my-dependency"` and using "Module" later, if that changes anything.

**Comment by dmnsgn** [comment]
Also when writing to fs (by not setting any compiler file system), no issue.

**Comment by privatenumber** [comment]
I can try to take a look if you post your configuration code.

**Comment by dmnsgn** [comment]
Thank you. I have put together a repro as a gist: https://gist.github.com/dmnsgn/4214e9419161c37372ede8c48045ecad

### Install it with:
```bash
git clone https://gist.github.com/4214e9419161c37372ede8c48045ecad.git
cd 4214e9419161c37372ede8c48045ecad/
npm i
```

### What it does:
- starts an express server (with webpackDevMiddleware)
- run webpack with `index.js` as entry point
- `index.js` is generated by the MyPlugin in memoryFs
- `index.js` imports all `dependencies` from `package.json` and logs them to console

### How to test the issue:
- `npm start` 
- Open http://localhost:8080 and check the console
- it should log only one module (primitive-geometry)
- while server still runs, install any dependency (`npm i cameras` for instance) then check again the console
-  package.json will get updated and you can log `indexDependencies` in generateIndex (or memfs vol.toJSON() to double check)

### Expected behaviour:
on refreshing browser window, console should now logs 2 dependencies (primitive-geometry and the other you just installed)

### Actual behaviour:
only the first one is still logged

### Notes:
Setting `cache: false,` works.
Writing `compiler.intermediateFileSystem` and `compiler.outputFileSystem` to regular file system works.
Changing webpackDevMiddleware `outputFileSystem` doesn't seem to have any effect on this.

**Comment by privatenumber** [comment]
As suspected, you didn't follow the specific recommendation [here](https://github.com/webpack/webpack/issues/13372#issuecomment-841598405).

This issue is about [`compiler.intermediateFileSystem = memoryFs`](https://gist.github.com/dmnsgn/4214e9419161c37372ede8c48045ecad#file-run-js-L107) not [working expectedly](https://github.com/webpack/webpack/issues/13372#issuecomment-840125734).

**Comment by dmnsgn** [comment]
I did try your suggestion of adding a plugin to setup file systems early but it had no effect: following the "How to test the issue" steps will still result in showing only one dependency being logged instead of two after reload.

If you want to double check, I have updated the gist with a `ConfigureCompilerPlugin` instance to confirm that: https://gist.github.com/dmnsgn/4214e9419161c37372ede8c48045ecad#file-run-js-L109-L113
(Also added `console.log(memoryFs.toJSON()[inMemoryIndex]);` to show that the file in memory is correctly updated)

**Comment by dmnsgn** [comment]
> This issue is about [`compiler.intermediateFileSystem = memoryFs`](https://gist.github.com/dmnsgn/4214e9419161c37372ede8c48045ecad#file-run-js-L107) not [working expectedly](https://github.com/webpack/webpack/issues/13372#issuecomment-840125734).

@privatenumber are you saying this should be a separate issue?

I would say that he fact it doesn't even work when setting `compiler.intermediateFileSystem` early (with a plugin) and that disabling cache (`{ cache: false }`) do work, points to an issue either with the webpack-dev-middleware or the webpack cache.

**Comment by dmnsgn** [comment]
@privatenumber Looking at it differently, is there a way to programatically invalidate the cache for a specific file? (in memory or not)

**Comment by dmnsgn** [comment]
More info:

- after installing a new dependency, I can see the entry point file is updated correctly `http://localhost:8080/index.js` when loaded directly in the browser
- looking at devtool `Sources` though (`webpack://webpack-memfs-cache/../../%00#test/4214e9419161c37372ede8c48045ecad/index.js?`), I see the webpack generated code isn't updated and only contain the previous state, with only one dependency: 
```js
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var canvas_screenshot__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! canvas-screenshot */ "../../\u0000#test/4214e9419161c37372ede8c48045ecad/node_modules/canvas-screenshot/index.js");

console.log(canvas_screenshot__WEBPACK_IMPORTED_MODULE_0__);
```

So I am deducing webpack is not picking up changes because it is cached and think it is unchanged somehow.

PS: same error with cache `type: "memory"`
